### PR TITLE
Git feature

### PR DESCRIPTION
**Pull request by cinghiopinghio**
Hi,

This is my attempt to implement the github feature as in #27 .

I added the `-g` command argument to use `git ls-files` as a file and folder factory.

It works in my environment and passed all test in your test folder (after adding `-g` to `stowsh` in all `run` scripts).

Yet to be done: add permanently tests to the test suite.

### CODE REVIEW

**Comment by mikepqr** [comment]
This looks great too! Thanks. I'll take a proper look as soon as I can.

In the meantime, the test system is very hacky but it should be self explanatory. You create a directory for the package and directories for what the target look like once it's installed and after it's uninstalled. In this case you'd presumably want to test .gitignored files and untracked files do not make it into the installation if `-g` is on. Test coverage isn't great (#21) and I have vague plans to move to a real bash unit test framework, so I won't block this merge on the absence of tests, but it would be great to see them!

**Comment by cinghiopinghio** [comment]
I have added a test for the git feature.
It was actually useful as I could spot a small bug in folders with no git files!
Cheers.

File: stowsh
Comment (mikepqr) [inline stowsh]: Unfortunately this doesn't work on macOS, where the BSD version of dirname doesn't accept more than one argument. If possible, I would prefer this to be rewritten as a loop or with xargs to accumulate the result of multiple calls. This would avoid adding a coreutils dependency (see #14).

But, if you can't figure it out ... stowsh already has GNU coreutils as a requirement, and relying on GNU behavior here wouldn't make the situation significantly worse than it already is. If you stick with GNU dirname then please follow the pattern in `stowsh_setpaths` and `isgnu` to set it up.

The general idea is that the script should first check for `g<utility>` and use that if it exists (because homebrew installs GNU coreutils with a `g` prefix by default). And then it should check `<utility>` is indeed the GNU version (and abort if not).

**Review by mikepqr — CHANGES_REQUESTED** [review]
Looks great! Just one issue that does need fixing, but should be pretty straightforward. Let me know if you have any trouble with my request.

File: stowsh
Comment (mikepqr) [reply stowsh]: If you go with `xargs` bear in mind that the BSD xargs on macOS doesn't support long GNU switches, but I think it does what you need, i.e.
```
$ paths=(foo/1 bar/2 baz/3)
$ printf "%s\n" ${paths[@]} | xargs -n1 dirname
foo
bar
baz
```

Comment (cinghiopinghio) [reply stowsh]: I will go the `xargs` way, just let me find 5 minutes to test and commit.

**Review by cinghiopinghio — COMMENTED** [review]
Should work now

**Comment by cinghiopinghio** [comment]
I'm not sure if I have any more tasks to do in this "review request" (first timer).

File: stowsh
Comment (mikepqr) [inline stowsh]: I think you need double quotes around `${gitfiles[@]}`.

**Review by mikepqr — CHANGES_REQUESTED** [review]
Thanks! LGTM. Tested on macOS.

I think this is ready to merge once you add the quotes (see inline comment) and squash  into a single commit. (I think the standard github workflow for that is something like `git rebase -i main` to tidy up, and then `git push -f` once you're done to update this PR. Let me know if you have any difficulties.)

File: stowsh
Comment (cinghiopinghio) [reply stowsh]: I think it's not necessary: `xargs` already picks one folder per line. Anyway, just to be on the safe part, I added the double quotes and squashed.

Comment (mikepqr) [reply stowsh]: Thanks. My general rule is: if [shellcheck](https://github.com/koalaman/shellcheck/) tells me to do something, I do it ;-)

**Comment by mikepqr** [comment]
Thank you for your contribution!

**Pull request merged**
